### PR TITLE
Log argument parsing improvements

### LINKED ISSUE #494: `kart log` doesn't handle `--` the same way as git
`kart log` passes args off to `git log`, _except_ that it parses them using click first. Since click swallows the `--` arg, that means it never makes it to git.

So:

* `kart log {datasetname}` --> `git log {datasetname}`
* `kart log -- {datasetname}` invokes `git log {datasetname}`
* `kart log -- -- {datasetname}` invokes `git log -- {datasetname}`

We should fix this.

The only _obvious_ way I can see to fix it is to inspect `sys.argv` before click gets to it, and if there's a `--`, just insert _another_ `--`. But we should probably figure out if there's a click way to do it

### PR DESCRIPTION
**Pull request by craigds**
## Description

Improves parsing of arguments in `kart log`

## Related links:

- Fixes #494 
- This PR is prerequisite work for #496 and #493

## Checklist:

- [x] Have you reviewed your own change?
- [x] Have you included test(s)?
- [x] Have you updated the [changelog](https://github.com/koordinates/kart/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: kart/log.py [hunk 1]
```diff
@@ -51,19 +85,62 @@
...
+    # Handle positional args.
+    # These could be refs, commits, trees, or paths.
+    #  * as soon as we encounter a path, we assume all remaining args are also paths.
+    #  * i.e. the paths must be given *last*.
+    #  * If it's ambiguous whether something is a path or not, we assume it's a commit-ish.
+    #  * If you want to be unambiguous, provide the `--` arg to separate the list of commit-ish-es and paths.
+    # This behaviour is consistent with git's behaviour.
+    try:
+        dash_index = args.index("--")
+    except ValueError:
```
Comment (olsen232) [inline kart/log.py:99]: Maybe it's pythonic how it is but I don't love try-except-else for readability - I kinda prefer:

try:
  dash_index = ...
except ValueError:
  dash_index = None
  
  if dash_index is None:
    ...

File: kart/log.py [hunk 2]
```diff
@@ -51,19 +85,62 @@
 @click.argument("args", nargs=-1, type=click.UNPROCESSED)
 def log(ctx, output_format, json_style, do_dataset_changes, with_feature_count, args):
     """ Show commit logs """
+    repo = ctx.obj.get_repo(allowed_states=KartRepoState.ALL_STATES)
+
+    # Handle positional args.
```
Comment (olsen232) [inline kart/log.py:90]: Arg parsing could probably have it's own function. No harm in adding more functions, makes code easier to read.

**Review by olsen232 — APPROVED** [review]

File: kart/log.py (hunk 1 above)
Comment (craigds) [reply kart/log.py:99]: I prefer the `else:` myself.

But now that I'm thinking about it, I actually prefer this to both of those:

```python

    if "--" in args:
        dash_index = args.index("--")
        paths = args[dash_index + 1 :]
        other_args = args[:dash_index]
    else:
        ...
```

**Pull request merged**